pandas/compat/_optional.py

```diff
@@ -0,0 +1,100 @@
+import distutils.version
+import importlib
+import warnings
+
+# Update install.rst when updating versions!
+
+VERSIONS = {
+    "bs4": "4.4.1",
+    "fastparquet": "0.2.1",
+    "gcsfs": "0.1.0",
+    "matplotlib": "2.2.2",
+    "numexpr": "2.6.2",
+    "openpyxl": "2.4.0",
+    "pandas_gbq": "0.8.0",
+    "pyarrow": "0.9.0",
+    "pytables": "3.4.2",
+    "s3fs": "0.0.8",
+    "scipy": "0.19.0",
+    "sqlalchemy": "1.1.4",
+    "xarray": "0.8.2",
+    "xlrd": "1.0.0",
+    "xlwt": "2.4.0",
+}
+
+message = (
+    "Missing optional dependency '{name}'. {extra} "
+    "Use pip or conda to install {name}."
+)
+version_message = (
+    "Pandas requires version '{minimum_version}' or newer of '{name}' "
+    "(version '{actual_version}' currently installed)."
+)
+
+
+def _get_version(module):
+    try:
+        version = module.__version__
+    except AttributeError:
+        # xlrd uses a capitalized attribute name
+        version = module.__VERSION__
+    return version
+
+
+def import_optional_dependency(
+    name, extra="", raise_on_missing=True, on_version="raise"
+):
+    """
+    Import an optional dependency.
+
+    By default, if a dependency is missing an ImportError with a nice
+    message will be raised. If a dependency is present, but too old,
+    we raise.
+
+    Parameters
+    ----------
+    name : str
+        The module name. This should be top-level only, so that the
+        version may be checked.
+    extra : str
+        Additional text to include in the ImportError message.
+    raise_on_missing : bool, default True
+        Whether to raise if the optional dependency is not found.
+        When False and the module is not present, None is returned.
+    on_version : str {'raise', 'warn'}
+        What to do when a dependency's version is too old.
+
+        * raise : raise an ImportError
+        * warn : Warn that the version is too old. Returns None
+
+    Returns
+    -------
+    maybe_module : Optional[ModuleType]
+        The imported module, when found and the version is correct.
+        None is returned when the package is not found and `raise_on_missing`
+        is False, or when the package's version is too old and `on_version`
+        is ``'warn'``.
+    """
+    try:
+        module = importlib.import_module(name)
+    except ImportError:
+        if raise_on_missing:
+            raise ImportError(message.format(name=name, extra=extra)) from None
+        else:
+            return
+
+    minimum_version = VERSIONS.get(name)
```

Comment:
We should consider setting a minimum version for every optional dep, and just making this a `__getitem__`

Reply:
I think you *do* this now? so can you change to ``[]``?

Reply:
There are ones, like Jinja2, where we don't specify a minimum supported version.